simln-lib/src/sim_node.rs

```diff
@@ -692,6 +704,178 @@ impl<T: SimNetwork> LightningNode for SimNode<'_, T> {
+    // of the htlc if another interceptor already signalled a failure.
+    pub shutdown_listener: Listener,
+}
+
+impl Display for InterceptRequest {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "htlc forwarded by {} over {}:{} -> {} forward amounts {} {}",
+            self.forwarding_node,
+            self.incoming_htlc.channel_id,
+            self.incoming_htlc.index,
+            {
+                if let Some(c) = self.outgoing_channel_id {
+                    format!("-> {c}")
+                } else {
+                    "receive".to_string()
+                }
+            },
+            self.incoming_amount_msat,
+            self.outgoing_amount_msat
+        )
+    }
+}
+
+/// Notification sent to an external interceptor notifying that a htlc that was previously intercepted has been
+/// resolved.
+pub struct InterceptResolution {
+    /// The node that is forwarding this HTLC.
+    pub forwarding_node: PublicKey,
+
+    /// Unique identifier for the incoming htlc.
+    pub incoming_htlc: HtlcRef,
+
+    /// The short channel id for the outgoing channel that this htlc should be forwarded over, None if notifying the
+    /// receiving node.
+    pub outgoing_channel_id: Option<ShortChannelID>,
+
+    /// True if the htlc was settled successfully.
+    pub success: bool,
+}
+
+// Custom records that can be attached by an Interceptor to intercepted HTLCs.
+pub type CustomRecords = HashMap<u64, Vec<u8>>;
+
+#[derive(Debug, Clone)]
+pub struct HtlcRef {
+    pub channel_id: ShortChannelID,
+    pub index: u64,
+}
+
+async fn handle_intercepted_htlc(
+    request: InterceptRequest,
+    interceptors: &[Arc<dyn Interceptor>],
+    trigger: Trigger,
+) -> Result<Result<CustomRecords, ForwardingError>, CriticalError> {
+    let mut attached_custom_records: CustomRecords = HashMap::new();
+    if !interceptors.is_empty() {
+        let mut intercepts: JoinSet<Result<Result<CustomRecords, ForwardingError>, CriticalError>> =
+            JoinSet::new();
+
+        for interceptor in interceptors.iter() {
+            let request = request.clone();
+            log::trace!(
+                "Sending HTLC to intercepor: {} {request}",
+                interceptor.name()
+            );
+
+            let interceptor_clone = Arc::clone(interceptor);
+            intercepts.spawn(async move { interceptor_clone.intercept_htlc(request).await });
+        }
+
+        // Read results from the interceptors and check whether any of them returned an instruction to fail
+        // the HTLC. If any of the interceptors did return an error, we send a shutdown signal
+        // to the other interceptors that may have not returned yet.
+        let mut interceptor_failure = None;
+        while let Some(res) = intercepts.join_next().await {
+            let res =
+                res.map_err(|join_err| CriticalError::InterceptorError(join_err.to_string()))?;
```

Comment:
Wondering about how best to handle `CriticalError` but here and below for `DuplicateCustomRecords`.

When we error out, will `shutdown_trigger` be called somewhere up the stack? Otherwise the other interceptors don't get the signal to shut down. 

I think that the strictly cleanest way to do this would be to:
1. Track that we ran into a critical error
2. Trigger shutdown
3. Loop through remaining tasks to make sure they completed
4. Return the first critical error we hit

Doesn't need to be in this PR, but maybe let's write this up in an issue just so that we don't forget?

Reply:
Yeah this is very nitpickey - in the case of critical shutdown a few interceptors might get aborted because the joinset is dropped (this will actually happen before shutdown is triggered, so my previous comment about it getting triggered up the stack doesn't really matter).

Let's leave as-is and make an issue noting that we don't guarantee clean shutdown of interceptors when we hit a `CriticalError`. I don't see it being much of a problem in practice, but good to have an issue so it's known at least.